tests/conftest.py

```diff
@@ -62,18 +75,405 @@ def conn_params(test_session):
+    return make_test_dfv_dsg_time_series(
+        1, file=data_file_1, variable_alias=variable_alias_1)
+
+
+@pytest.fixture(scope='function')
+def dfv_dsg_time_series_12(data_file_1, variable_alias_2):
+    return make_test_dfv_dsg_time_series(
+        2, file=data_file_1, variable_alias=variable_alias_2)
+
+
+@pytest.fixture(scope='function')
+def dfv_dsg_time_series_21(data_file_2, variable_alias_1):
+    return make_test_dfv_dsg_time_series(
+        3, file=data_file_2, variable_alias=variable_alias_1)
+
+
+@pytest.fixture(scope='function')
+def dfv_dsg_time_series_31(data_file_3, variable_alias_1):
+    return make_test_dfv_dsg_time_series(
+        4, file=data_file_3, variable_alias=variable_alias_1)
+
+
+# Ensemble
+
+def make_ensemble(id):
+    return Ensemble(
+        id=id,
+        changes='wonder what this is for',
+        description='Ensemble {}'.format(id),
+        name='ensemble_{}'.format(id),
+        version=float(id)
+    )
+
+
+@pytest.fixture(scope='function')
+def ensemble_1():
+    return make_ensemble(1)
+
+
+@pytest.fixture(scope='function')
+def ensemble_2():
+    return make_ensemble(2)
+
+# EnsembleDataFileVariables: tag DFVs with Ensembles
+
+def make_ensemble_dfvs(ensemble, dfv):
+    return EnsembleDataFileVariables(
+        ensemble_id=ensemble.id,
+        data_file_variable_id=dfv.id,
+    )
+
+
+@pytest.fixture(scope='function')
+def ensemble_dfvs_1(ensemble_1, dfv_dsg_time_series_11, dfv_dsg_time_series_21):
+    return [
+        make_ensemble_dfvs(ensemble_1, dfv_dsg_time_series_11),
+        make_ensemble_dfvs(ensemble_1, dfv_dsg_time_series_21),
+    ]
+
+
+# TODO: Make this, or something like it, drive ensemble_dfvs_1
+@pytest.fixture(scope='function')
+def ensemble_1_data_files(data_file_1, data_file_2):
+    return {data_file_1, data_file_2}
+
+
+@pytest.fixture(scope='function')
+def ensemble_dfvs_2(ensemble_2, dfv_dsg_time_series_21, dfv_dsg_time_series_31):
+    return [
+        make_ensemble_dfvs(ensemble_2, dfv_dsg_time_series_21),
+        make_ensemble_dfvs(ensemble_2, dfv_dsg_time_series_31),
+    ]
+
+
+# TODO: Make this, or something like it, drive ensemble_dfvs_2
+@pytest.fixture(scope='function')
+def ensemble_2_data_files(data_file_3, data_file_2):
+    return {data_file_2, data_file_3}
+
+
+# Database sessions
+
+@pytest.fixture(scope="function")
+def mm_test_session_objects(
+    model_1,
+    emission_1,
+    emission_2,
+    run_11,
+    run_12,
+    data_file_1,
+    data_file_2,
+    data_file_3,
+    dfv_dsg_time_series_11,
+    dfv_dsg_time_series_12,
+    dfv_dsg_time_series_21,
+    dfv_dsg_time_series_31,
+    ensemble_1,
+    ensemble_2,
+    ensemble_dfvs_1,
+    ensemble_dfvs_2,
+):
+    # Note: Order matters. These objects must be inserted in the order given,
+    # and removed in the reverse order.
+    # Note: Completeness matters. SQLAlchemy will implicitly add objects but
+    # not necessarily implicitly delete them.
```

Comment:
IIRC correctly, this depends on some configuration in the relationship definitions and can be changed.

Reply:
Yes, and,
- only for cascadable deletes; many-to-many relationships I don't think are in this category
- that would require updating modelmeta with quite a few new or revised relationships -- which I'd like to do, but now?